# Capstone Project - The Battle of the Neighborhoods (Week 2)
### Applied Data Science Capstone by IBM/Coursera

## <span style="color:darkred">Table of contents</span>
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussions](#results)
* [Conclusion](#conclusion)

## <span style="color:darkred"> 1. Introduction: Business Problem <a name="introduction"></a></span>

In this project, we will investigate the influence of poverty rate to the facilities and amenities of certain borough and how this insights can provide targeting local and business development for London boroughs. It will be interesting to see how poverty impacts the facilities, venues and amenities available in the boroughs. 

<br><br>
<center><img src="poverty_rate.jpg"
     alt="London Poverty Rates"
     width = "700"
     style="float: centre; margin-left: 10px;" /></center><br><br>


This will be done by carrying out a comparative studies on the facilities and amenities between the two boroughs with the highest income rate and the lowest in London, **Tower Hamlet (TH)** borough and **Bromley (Brom)** borough of London, as shown in the figure above.  

We will use the data science principles and techniques learned to generate a model of these boroughs, looking at the venues and facilities available in these areas and provides insights to stakeholders, the local authorities and business chambers of commerce.

### Questions to answer:

1.	What types of facilities (venues) and amenities are available in the wards (neighbourhoods) with different poverty line?
2.	How venues changing based on the spending power?
3.	What are the distinctive venues that represent in these boroughs?
4.	Suggestions and recommendations. 

By answering the above questions, the findings can be used for the targeting development for the rest of the London Boroughs so that unnecessary development can be avoided and overall budget can be sustained.

======================================================================================================================== 

## <span style="color:darkred">2. Data <a name="data"></a></span>

Based on the problem definition, factors that will influence the decision in this project will be:

* number and the type of venues and facilities available in the suurounding area of these boroughs
* the most frequent venues and facilties for each boroughs

To define the surrounding area of the borough, we will be using:
* London Boroughs poverty information from: **https://www.trustforlondon.org.uk/data/poverty-borough**. *Accessed: 11/03/2019*
* latitudes and longitudes of Tower Hamlet and Bromley  obtained from: **https://www.distancesto.com/coordinates/gb/**. *Accessed: 11/03/2019*  
* venues, type and locations in every borough will be obtained using **Foursquare API**

Further information of the boroughs can be found at:
* Bromley: **https://en.wikipedia.org/wiki/London_Borough_of_Bromley** 
* Tower Hamlet: **https://en.wikipedia.org/wiki/London_Borough_of_Tower_Hamlets**

## Boroughs Locations 
Based on the information obtained from https://www.distancesto.com/coordinates/gb/, the latitude and the longitude of TM and Brom are as follows:

In [1]:
TH_coordinates = (51.520261, -0.02934)
BROM_coordinates = (51.367971, 0.070062)
london_coordinates = (51.509865, -0.118092)

Let's visualise the locations of these boroughs on London map:

In [2]:
import folium
from folium.features import DivIcon

london_map = folium.Map(location = london_coordinates, zoom_start = 10)

folium.Circle(
    radius=2500,   # the radius is calculated based on the area coverage of the borough 
    location= TH_coordinates,
    color= 'crimson',
    fill=False,
).add_to(london_map)

folium.Marker(
    TH_coordinates, 
    popup=('Tower Hamlet'), 
    icon=folium.Icon(color='crimson', 
    icon_color='white', icon='info-sign', angle=0, prefix='fa')
).add_to(london_map)

folium.Circle(
    radius=6900, # the radius is calculated based on the area coverage of the borough 
    location= BROM_coordinates, 
    popup = 'Bromley',
    color='darkblue',
    fill=False,
).add_to(london_map)

folium.Marker(
    BROM_coordinates, 
    popup=('Bromley'), 
    icon=folium.Icon(color='darkblue', 
    icon_color='white', icon='info-sign', angle=0, prefix='fa')
).add_to(london_map)



london_map

Preliminary observations from the above map show that based on the locations for the two boroughs, Tower Hamlet is located very near to the London centre, where as Bromley borough located at the boundary of the M25 Ringroad, which is about 2 hours drive from London centre. In terms of area size, Tower Hamlet convers about 19.77km2 and Bromley is about 150.2km2. Based on these information, we can used them as distance references when retrieving venues, type and locations  using Foursquare API.

Now that we have the information about the boroughs, let's load the wards (neighbourhood) information of each boroughs.

In [3]:
import pandas as pd

TH_data = pd.read_csv('TH_neighbourhoods.csv')
TH_data.head()

neighbourhood   latitude  longitude
0              Bethnal Green  51.526962  -0.066740
1  Blackwall and Cubitt Town  51.495182  -0.009826
2                   Bow East  51.528309  -0.019482
3                   Bow West  51.528309  -0.019482
4               Canary Wharf  51.505219  -0.018900

In [4]:
print('Borough of Tower Hamlet has {} neighbourhoods.'.format(
        len(TH_data['neighbourhood'].unique()),
        TH_data.shape[0]
    )
)

Borough of Tower Hamlet has 15 neighbourhoods.


In [5]:
BROM_data = pd.read_csv('BROM_neighbourhoods.csv')
BROM_data.head()

neighbourhood   latitude  longitude
0                     Bickley  51.401740   0.043712
1                 Biggin Hill  51.331959   0.029057
2     Bromley Common & Keston  51.375875   0.043819
3                Bromley Town  51.402805   0.014814
4  Chelsfield & Pratts Bottom  51.357943   0.127288

In [6]:
print('Borough of Bromley has {} neighborhoods.'.format(
        len(BROM_data['neighbourhood'].unique()),
        BROM_data.shape[0]
    )
)

Borough of Bromley has 20 neighborhoods.


Now, let's confirm the locations of all the wards within the borough on the map based on the area coverage we have define earlier. b

In [7]:
london_map = folium.Map(location = london_coordinates, zoom_start = 10)

folium.Circle(
    radius=2500,   # the radius is calculated based on the area coverage of the borough 
    location= TH_coordinates,
    color= 'crimson',
    fill=False,
).add_to(london_map)

# add markers to map
for lat, lng, label in zip(TH_data['latitude'], TH_data['longitude'], TH_data['neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(london_map) 

folium.Circle(
    radius=6900, # the radius is calculated based on the area coverage of the borough 
    location= BROM_coordinates, 
    popup = 'Bromley',
    color='darkblue',
    fill=False,
).add_to(london_map)

# add markers to map
for lat, lng, label in zip(BROM_data['latitude'], BROM_data['longitude'], BROM_data['neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(london_map) 


london_map

## Foursquare API

Now that we have our location candidates, let's use Foursquare API to get info on venues in each of the wards within each borough

As an exploratory project, we will retrieve the venues based on the areas of each borough. We will then do the neccesary manipulations and analysis to achieve our objectives.  

### Define Foursquare Credential and Version 

In [8]:
CLIENT_ID = 'VB4GSHAOKEPLPPVS0VBVRAXL3DXVHHTRU3BJ4X4NJSGSF3R4' # your Foursquare ID
CLIENT_SECRET = '0N21QNS5SLASFGZFAGYFVAGKCE15NM40N2ZOXLAHS50KHICP' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VB4GSHAOKEPLPPVS0VBVRAXL3DXVHHTRU3BJ4X4NJSGSF3R4
CLIENT_SECRET:0N21QNS5SLASFGZFAGYFVAGKCE15NM40N2ZOXLAHS50KHICP


In [9]:
# importing the neccesary libraries for the tasks 

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
import pandas as pd
import numpy as np
import json # library to handle JSON files
import requests # library to handle requests

# import k-means from clustering stage
from sklearn.cluster import KMeans

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
import matplotlib.pyplot as plt
%matplotlib inline 

In [10]:
# fundtion to repeat the same process of retrieving venues of all the neighbourhoods 

def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT = 100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Tower Hamlet: Let's get the venues in Tower Hamlet within a radius of 500 meters of each neighbourhoods

In [11]:
TH_venues = getNearbyVenues(names=TH_data['neighbourhood'],
                                   latitudes = TH_data['latitude'],
                                   longitudes = TH_data['longitude']
                                  )



Bethnal Green
Blackwall and Cubitt Town
Bow East
Bow West
Canary Wharf
Island Gardens 
Lansbury
Limehouse 
Poplar
Shadwell
Spitalfields and Banglatown
St Katharine's and Wapping 
Stepney Green
Weavers
Whitechapel


In [12]:
print(TH_venues.shape)

TH_venues.to_csv('TH_venues', index = False)
TH_venues.head()

(520, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Bethnal Green               51.526962                 -0.06674   
1  Bethnal Green               51.526962                 -0.06674   
2  Bethnal Green               51.526962                 -0.06674   
3  Bethnal Green               51.526962                 -0.06674   
4  Bethnal Green               51.526962                 -0.06674   

             Venue  Venue Latitude  Venue Longitude Venue Category  
0  The King's Arms       51.525754        -0.065868            Pub  
1       Sam's Cafe       51.526424        -0.065056           Café  
2       Woolidando       51.526377        -0.066518           Café  
3        Jonestown       51.526092        -0.067936    Coffee Shop  
4       E Pellicci       51.526516        -0.063426           Café

In [13]:
TH_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  Neighbourhood Longitude  \
Neighbourhood                                                                  
Bethnal Green                                    63                       63   
Blackwall and Cubitt Town                        32                       32   
Bow East                                         13                       13   
Bow West                                         13                       13   
Canary Wharf                                    100                      100   
Island Gardens                                   30                       30   
Lansbury                                         16                       16   
Limehouse                                        28                       28   
Poplar                                           13                       13   
Shadwell                                         22                       22   
Spitalfields and Banglatown                     100                      100   
St Katharine's and Wapping                       15                       15   
Stepney Green                                    20                       20   
Weavers                                          21                       21   
Whitechapel                                      34                       34   

                             Venue  Venue Latitude  Venue Longitude  \
Neighbourhood                                                         
Bethnal Green                   63              63               63   
Blackwall and Cubitt Town       32              32               32   
Bow East                        13              13               13   
Bow West                        13              13               13   
Canary Wharf                   100             100              100   
Island Gardens                  30              30               30   
Lansbury                        16              16               16   
Limehouse                       28              28               28   
Poplar                          13              13               13   
Shadwell                        22              22               22   
Spitalfields and Banglatown    100             100              100   
St Katharine's and Wapping      15              15               15   
Stepney Green                   20              20               20   
Weavers                         21              21               21   
Whitechapel                     34              34               34   

                             Venue Category  
Neighbourhood                                
Bethnal Green                            63  
Blackwall and Cubitt Town                32  
Bow East                                 13  
Bow West                                 13  
Canary Wharf                            100  
Island Gardens                           30  
Lansbury                                 16  
Limehouse                                28  
Poplar                                   13  
Shadwell                                 22  
Spitalfields and Banglatown             100  
St Katharine's and Wapping               15  
Stepney Green                            20  
Weavers                                  21  
Whitechapel                              34

#### Let's find out how many unique categories can we obtained from Tower Hamlet

In [14]:
print('There are {} uniques venue categories at Tower Hamlet.'.format(len(TH_venues['Venue Category'].unique())))

There are 156 uniques venue categories at Tower Hamlet.


### Bromley: Let's get the venues in Bromley within a radius of 500 meters of each neighbourhoods

In [15]:
BROM_venues = getNearbyVenues(names=BROM_data['neighbourhood'],
                                   latitudes = BROM_data['latitude'],
                                   longitudes = BROM_data['longitude']
                                  )



Bickley
Biggin Hill
Bromley Common & Keston
Bromley Town
Chelsfield & Pratts Bottom
Chislehurst
Copers Cope
Cray Valley East
Crystal Palace
Darwin
Farnborough & Croftton
Hayes & Coney Hall
Kelsey & Eden Park
Mottingham & Chislehurst North
Orpington
Penge & Cator
Petts Wood & Knoll
Plaistow & Sunbridge 
Shortlangs 
West Wickham 


In [16]:
print(BROM_venues.shape)

BROM_venues.to_csv('BROM_venues', index = False)
BROM_venues.head()

(193, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0       Bickley               51.401740                 0.043712   
1       Bickley               51.401740                 0.043712   
2       Bickley               51.401740                 0.043712   
3       Bickley               51.401740                 0.043712   
4   Biggin Hill               51.331959                 0.029057   

                                             Venue  Venue Latitude  \
0                    Bickley Railway Station (BKL)       51.400032   
1                        Bickley Park Cricket Club       51.401508   
2                                 J Henry Flooring       51.401227   
3                             Village Sandwich Bar       51.399301   
4  Biggin Hill Airport (BQH) (Biggin Hill Airport)       51.331794   

   Venue Longitude  Venue Category  
0         0.045351   Train Station  
1         0.046263  Cricket Ground  
2         0.040652    Home Service  
3         0.047951            Café  
4         0.028845         Airport

In [17]:
BROM_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                            
Bickley                                              4   
Biggin Hill                                          4   
Bromley Common & Keston                              5   
Bromley Town                                        45   
Chelsfield & Pratts Bottom                           3   
Chislehurst                                          6   
Copers Cope                                          4   
Cray Valley East                                    28   
Crystal Palace                                      22   
Darwin                                               3   
Farnborough & Croftton                               5   
Hayes & Coney Hall                                   5   
Kelsey & Eden Park                                   4   
Mottingham & Chislehurst North                       4   
Orpington                                           19   
Penge & Cator                                       17   
Petts Wood & Knoll                                   3   
Plaistow & Sunbridge                                 4   
Shortlangs                                           5   
West Wickham                                         3   

                                Neighbourhood Longitude  Venue  \
Neighbourhood                                                    
Bickley                                               4      4   
Biggin Hill                                           4      4   
Bromley Common & Keston                               5      5   
Bromley Town                                         45     45   
Chelsfield & Pratts Bottom                            3      3   
Chislehurst                                           6      6   
Copers Cope                                           4      4   
Cray Valley East                                     28     28   
Crystal Palace                                       22     22   
Darwin                                                3      3   
Farnborough & Croftton                                5      5   
Hayes & Coney Hall                                    5      5   
Kelsey & Eden Park                                    4      4   
Mottingham & Chislehurst North                        4      4   
Orpington                                            19     19   
Penge & Cator                                        17     17   
Petts Wood & Knoll                                    3      3   
Plaistow & Sunbridge                                  4      4   
Shortlangs                                            5      5   
West Wickham                                          3      3   

                                Venue Latitude  Venue Longitude  \
Neighbourhood                                                     
Bickley                                      4                4   
Biggin Hill                                  4                4   
Bromley Common & Keston                      5                5   
Bromley Town                                45               45   
Chelsfield & Pratts Bottom                   3                3   
Chislehurst                                  6                6   
Copers Cope                                  4                4   
Cray Valley East                            28               28   
Crystal Palace                              22               22   
Darwin                                       3                3   
Farnborough & Croftton                       5                5   
Hayes & Coney Hall                           5                5   
Kelsey & Eden Park                           4                4   
Mottingham & Chislehurst North               4                4   
Orpington                                   19               19   
Penge & Cator                               17               17   
Petts Wood & Knoll                           3                3   
Plaistow & Sunbridge         

In [18]:
print('There are {} uniques venue categories at Bromley.'.format(len(BROM_venues['Venue Category'].unique())))

There are 76 uniques venue categories at Bromley.


####  In summary:
* Total venues in Tower Hamlet neighbourhoods returned by Foursquare: **520**
* Total venues in Bromley neighbourhoods returned by Foursquare: **193**
* Total unique venue categories in Tower Hamlet = **156**
* Total unique venue categories in Bromley = **76**
* Tower Hamlet venues dataframe is called: **TH_venues**
* Tower Hamlet venues dataframe is called: **BROM_venues**

## <span style="color:darkred">3. Methodology <a name="methodology"></a></span>

In this project, we will be only concentrating the two boroughs with the largest poverty rate gaps. 

In the first step, we will be looking into the top 5 most common venues for each of the borough, this will provides us with an overview of the types of venues popular in the boroughs. 

Second steps we will be clustering the neighbourhood for each borough to investigate the cluster formations 

in third and final steps, we will be drill into each of cluster to obtained the reviews of some of the venues to seek the quality of service at venues in this separate borough.

With these analysis, we will be able to drawn some conclusions on how poverty rate in particular borough impacting the venues in the areas using Foursquare geospatial data.

======================================================================================================================== 

## <span style="color:darkred">4. Analysis <a name="analysis"></a></span>

## Analyse Each Borough:

### Tower Hamlet:

In [19]:
# one hot encoding
TH_onehot = pd.get_dummies(TH_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
TH_onehot['Neighbourhood'] = TH_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [TH_onehot.columns[-1]] + list(TH_onehot.columns[:-1])
TH_onehot = TH_onehot[fixed_columns]

TH_onehot.head()

Neighbourhood  Art Gallery  Asian Restaurant  Athletics & Sports  \
0  Bethnal Green            0                 0                   0   
1  Bethnal Green            0                 0                   0   
2  Bethnal Green            0                 0                   0   
3  Bethnal Green            0                 0                   0   
4  Bethnal Green            0                 0                   0   

   BBQ Joint  Bagel Shop  Bakery  Bar  Beach  Beer Bar     ...       Trail  \
0          0           0       0    0      0         0     ...           0   
1          0           0       0    0      0         0     ...           0   
2          0           0       0    0      0         0     ...           0   
3          0           0       0    0      0         0     ...           0   
4          0           0       0    0      0         0     ...           0   

   Train Station  Tunnel  Turkish Restaurant  Vegetarian / Vegan Restaurant  \
0              0       0                   0                              0   
1              0       0                   0                              0   
2              0       0                   0                              0   
3              0       0                   0                              0   
4              0       0                   0                              0   

   Vietnamese Restaurant  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                      0         0            0              0            0  
1                      0         0            0              0            0  
2                      0         0            0              0            0  
3                      0         0            0              0            0  
4                      0         0            0              0            0  

[5 rows x 157 columns]

### Bromley:

In [20]:
# one hot encoding
BROM_onehot = pd.get_dummies(BROM_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
BROM_onehot['Neighbourhood'] = BROM_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [BROM_onehot.columns[-1]] + list(BROM_onehot.columns[:-1])
BROM_onehot = BROM_onehot[fixed_columns]

BROM_onehot.head()

Neighbourhood  Airport  Airport Service  American Restaurant  \
0       Bickley        0                0                    0   
1       Bickley        0                0                    0   
2       Bickley        0                0                    0   
3       Bickley        0                0                    0   
4   Biggin Hill        1                0                    0   

   Asian Restaurant  Athletics & Sports  Bakery  Bar  Bike Shop  Bookstore  \
0                 0                   0       0    0          0          0   
1                 0                   0       0    0          0          0   
2                 0                   0       0    0          0          0   
3                 0                   0       0    0          0          0   
4                 0                   0       0    0          0          0   

       ...        Stationery Store  Supermarket  Sushi Restaurant  \
0      ...                       0            0                 0   
1      ...                       0            0                 0   
2      ...                       0            0                 0   
3      ...                       0            0                 0   
4      ...                       0            0                 0   

   Tennis Court  Theater  Track Stadium  Train Station  Turkish Restaurant  \
0             0        0              0              1                   0   
1             0        0              0              0                   0   
2             0        0              0              0                   0   
3             0        0              0              0                   0   
4             0        0              0              0                   0   

   Wine Shop  Women's Store  
0          0              0  
1          0              0  
2          0              0  
3          0              0  
4          0              0  

[5 rows x 77 columns]

#### Next, let's group rows by neighbourhood and by taking the mean of the frequency of occurrence of each category|

In [21]:
TH_grouped = TH_onehot.groupby('Neighbourhood').mean().reset_index()
TH_grouped

Neighbourhood  Art Gallery  Asian Restaurant  \
0                 Bethnal Green     0.015873            0.0000   
1     Blackwall and Cubitt Town     0.000000            0.0000   
2                      Bow East     0.076923            0.0000   
3                      Bow West     0.076923            0.0000   
4                  Canary Wharf     0.000000            0.0100   
5               Island Gardens      0.033333            0.0000   
6                      Lansbury     0.000000            0.0625   
7                    Limehouse      0.000000            0.0000   
8                        Poplar     0.000000            0.0000   
9                      Shadwell     0.000000            0.0000   
10  Spitalfields and Banglatown     0.030000            0.0100   
11  St Katharine's and Wapping      0.000000            0.0000   
12                Stepney Green     0.000000            0.0000   
13                      Weavers     0.000000            0.0000   
14                  Whitechapel     0.000000            0.0000   

    Athletics & Sports  BBQ Joint  Bagel Shop    Bakery       Bar    Beach  \
0             0.000000       0.00    0.031746  0.015873  0.000000  0.00000   
1             0.000000       0.00    0.000000  0.000000  0.000000  0.03125   
2             0.000000       0.00    0.000000  0.000000  0.076923  0.00000   
3             0.000000       0.00    0.000000  0.000000  0.076923  0.00000   
4             0.000000       0.01    0.000000  0.030000  0.010000  0.00000   
5             0.033333       0.00    0.000000  0.000000  0.000000  0.00000   
6             0.000000       0.00    0.000000  0.000000  0.000000  0.00000   
7             0.035714       0.00    0.000000  0.000000  0.000000  0.00000   
8             0.000000       0.00    0.000000  0.000000  0.000000  0.00000   
9             0.000000       0.00    0.000000  0.000000  0.045455  0.00000   
10            0.000000       0.02    0.020000  0.000000  0.020000  0.00000   
11            0.000000       0.00    0.000000  0.000000  0.066667  0.00000   
12            0.000000       0.00    0.000000  0.050000  0.050000  0.00000   
13            0.000000       0.00    0.000000  0.047619  0.000000  0.00000   
14            0.000000       0.00    0.000000  0.029412  0.000000  0.00000   

    Beer Bar     ...          Trail  Train Station    Tunnel  \
0   0.015873     ...       0.000000       0.000000  0.000000   
1   0.000000     ...       0.000000       0.000000  0.000000   
2   0.000000     ...       0.000000       0.000000  0.000000   
3   0.000000     ...       0.000000       0.000000  0.000000   
4   0.010000     ...       0.000000       0.000000  0.000000   
5   0.000000     ...       0.033333       0.000000  0.033333   
6   0.000000     ...       0.000000       0.000000  0.000000   
7   0.000000     ...       0.000000       0.035714  0.000000   
8   0.000000     ...       0.000000       0.000000  0.076923   
9   0.000000     ...       0.000000       0.000000  0.000000   
10  0.020000     ...       0.000000       0.000000  0.000000   
11  0.000000     ...       0.000000       0.000000  0.000000   
12  0.000000     ...       0.000000       0.000000  0.000000   
13  0.000000     ...       0.000000       0.000000  0.000000   
14  0.000000     ...       0.000000       0.000000  0.000000   

    Turkish Restaurant  Vegetarian / Vegan Restaurant  Vietnamese Restaurant  \
0             0.015873                           0.00                   0.00   
1             0.000000                           0.00                   0.00   
2             0.000000                           0.00                   0.00   
3             0.000000                           0.00                   0.00   
4             0.010000                           0.00                   0.01   
5             0.000000                           0.00                   0.00   
6             0.000000                           0.00                   0.00   
7             0.035714                         

In [22]:
BROM_grouped = BROM_onehot.groupby('Neighbourhood').mean().reset_index()
BROM_grouped

Neighbourhood  Airport  Airport Service  \
0                          Bickley     0.00             0.00   
1                      Biggin Hill     0.25             0.25   
2          Bromley Common & Keston     0.00             0.00   
3                     Bromley Town     0.00             0.00   
4       Chelsfield & Pratts Bottom     0.00             0.00   
5                      Chislehurst     0.00             0.00   
6                      Copers Cope     0.00             0.00   
7                 Cray Valley East     0.00             0.00   
8                   Crystal Palace     0.00             0.00   
9                           Darwin     0.00             0.00   
10          Farnborough & Croftton     0.00             0.00   
11              Hayes & Coney Hall     0.00             0.00   
12              Kelsey & Eden Park     0.00             0.00   
13  Mottingham & Chislehurst North     0.00             0.00   
14                       Orpington     0.00             0.00   
15                   Penge & Cator     0.00             0.00   
16              Petts Wood & Knoll     0.00             0.00   
17           Plaistow & Sunbridge      0.00             0.00   
18                     Shortlangs      0.00             0.00   
19                   West Wickham      0.00             0.00   

    American Restaurant  Asian Restaurant  Athletics & Sports    Bakery  \
0                   0.0          0.000000            0.000000  0.000000   
1                   0.0          0.000000            0.000000  0.000000   
2                   0.0          0.000000            0.000000  0.000000   
3                   0.0          0.022222            0.000000  0.022222   
4                   0.0          0.000000            0.000000  0.000000   
5                   0.0          0.000000            0.000000  0.000000   
6                   0.0          0.000000            0.500000  0.000000   
7                   0.0          0.035714            0.000000  0.035714   
8                   0.0          0.000000            0.045455  0.000000   
9                   0.0          0.000000            0.000000  0.000000   
10                  0.2          0.000000            0.000000  0.000000   
11                  0.0          0.000000            0.000000  0.000000   
12                  0.0          0.000000            0.000000  0.000000   
13                  0.0          0.000000            0.000000  0.000000   
14                  0.0          0.000000            0.000000  0.052632   
15                  0.0          0.000000            0.000000  0.000000   
16                  0.0          0.000000            0.000000  0.000000   
17                  0.0          0.000000            0.000000  0.000000   
18                  0.0          0.000000            0.000000  0.000000   
19                  0.0          0.000000            0.000000  0.000000   

         Bar  Bike Shop  Bookstore      ...        Stationery Store  \
0   0.000000   0.000000   0.000000      ...                0.000000   
1   0.000000   0.000000   0.000000      ...                0.000000   
2   0.200000   0.000000   0.000000      ...                0.000000   
3   0.022222   0.000000   0.022222      ...                0.022222   
4   0.000000   0.000000   0.000000      ...                0.000000   
5   0.000000   0.000000   0.000000      ...                0.000000   
6   0.000000   0.000000   0.000000      ...                0.000000   
7   0.000000   0.000000   0.035714      ...                0.035714   
8   0.000000   0.045455   0.000000      ...                0.000000   
9   0.333333   0.000000   0.000000      ...                0.000000   
10  0.000000   0.000000   0.000000      ...                0.000000   
11  0.000000   0.000000   0.000000      ...                0.000000   
12  0.000000   0.000000   0.000000      ...                0.000000   
13  0.000000   0.000000   0.000000      ...                0.000000   
14  0.000000   0.000000   0.052632 

#### Let's print each Tower Hamlet neighborhood along with the top 5 most common venues

In [23]:
num_top_venues = 5

for hood in TH_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = TH_grouped[TH_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bethnal Green----
         venue  freq
0  Coffee Shop  0.13
1          Pub  0.11
2         Café  0.06
3         Park  0.03
4   Bagel Shop  0.03


----Blackwall and Cubitt Town----
               venue  freq
0  Indian Restaurant  0.09
1               Park  0.09
2               Café  0.06
3    Harbor / Marina  0.06
4      Grocery Store  0.06


----Bow East----
                      venue  freq
0                       Pub  0.15
1  Bike Rental / Bike Share  0.08
2      Fast Food Restaurant  0.08
3         Convenience Store  0.08
4               Coffee Shop  0.08


----Bow West----
                      venue  freq
0                       Pub  0.15
1  Bike Rental / Bike Share  0.08
2      Fast Food Restaurant  0.08
3         Convenience Store  0.08
4               Coffee Shop  0.08


----Canary Wharf----
                venue  freq
0         Coffee Shop  0.09
1        Burger Joint  0.05
2      Sandwich Place  0.05
3               Plaza  0.04
4  Italian Restaurant  0.03


----Island Gard

#### Let's print each Bromley neighborhood along with the top 5 most common venues

In [24]:
for hood in BROM_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = BROM_grouped[BROM_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bickley----
            venue  freq
0    Home Service  0.25
1   Train Station  0.25
2            Café  0.25
3  Cricket Ground  0.25
4         Airport  0.00


----Biggin Hill----
                venue  freq
0             Airport  0.25
1                 Pub  0.25
2      Massage Studio  0.25
3     Airport Service  0.25
4  Mexican Restaurant  0.00


----Bromley Common & Keston----
                  venue  freq
0           Bus Station   0.2
1                   Pub   0.2
2     Indian Restaurant   0.2
3                   Bar   0.2
4  Fast Food Restaurant   0.2


----Bromley Town----
                  venue  freq
0           Coffee Shop  0.13
1        Clothing Store  0.11
2  Gym / Fitness Center  0.07
3                   Pub  0.07
4           Pizza Place  0.04


----Chelsfield & Pratts Bottom----
                     venue  freq
0  Health & Beauty Service  0.33
1                      Pub  0.33
2            Fishing Store  0.33
3             Home Service  0.00
4           Ice Cream Shop  0.0

Function to sort the venues in descending order.

In [25]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

========================================================================================================================

### Top 10 venues for each neighborhood.

### Tower Hamlet:

In [26]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
TH_neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
TH_neighborhoods_venues_sorted['Neighbourhood'] = TH_grouped['Neighbourhood']

for ind in np.arange(TH_grouped.shape[0]):
    TH_neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(TH_grouped.iloc[ind, :], num_top_venues)


### Bromley:

In [27]:
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
BROM_neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
BROM_neighborhoods_venues_sorted['Neighbourhood'] = BROM_grouped['Neighbourhood']

for ind in np.arange(BROM_grouped.shape[0]):
    BROM_neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(BROM_grouped.iloc[ind, :], num_top_venues)


### Top 10 venues for each neighborhood.

In [28]:
TH_neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                 Bethnal Green           Coffee Shop   
1     Blackwall and Cubitt Town     Indian Restaurant   
2                      Bow East                   Pub   
3                      Bow West                   Pub   
4                  Canary Wharf           Coffee Shop   
5               Island Gardens                    Pub   
6                      Lansbury     Indian Restaurant   
7                    Limehouse                    Pub   
8                        Poplar                  Park   
9                      Shadwell         Grocery Store   
10  Spitalfields and Banglatown           Coffee Shop   
11  St Katharine's and Wapping                    Pub   
12                Stepney Green                  Park   
13                      Weavers                   Pub   
14                  Whitechapel                   Pub   

       2nd Most Common Venue          3rd Most Common Venue  \
0                        Pub                           Café   
1                       Park                Harbor / Marina   
2   Bike Rental / Bike Share                          Hotel   
3   Bike Rental / Bike Share                          Hotel   
4               Burger Joint                 Sandwich Place   
5             History Museum                  Boat or Ferry   
6              Grocery Store                    Beer Garden   
7         Chinese Restaurant              Indian Restaurant   
8        Fried Chicken Joint                  Grocery Store   
9                      Hotel                    Pizza Place   
10                       Pub  Vegetarian / Vegan Restaurant   
11                  Bus Stop                  Metro Station   
12                       Pub             Chinese Restaurant   
13               Supermarket          Portuguese Restaurant   
14                     Hotel                 Sandwich Place   

     4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                     Park            Bagel Shop            Restaurant   
1              Pizza Place         Grocery Store                   Pub   
2                      Gym         Metro Station                Office   
3                      Gym         Metro Station                Office   
4                    Plaza                  Café      Sushi Restaurant   
5                     Park           Art Gallery                Church   
6                     Café     Food & Drink Shop  Pakistani Restaurant   
7                     Café    Italian Restaurant           Pizza Place   
8       Chinese Restaurant            Steakhouse           Coffee Shop   
9                 Dive Bar          Burger Joint  Fast Food Restaurant   
10              Food Truck           Art Gallery     Indian Restaurant   
11      Chinese Restaurant    Persian Restaurant                   Bar   
12     Fried Chicken Joint       Thai Restaurant                  Farm   
13  Thrift / Vintage Store         Movie Theater        Sandwich Place   
14             Coffee Shop     Indian Restaurant        Ice Cream Shop   

        7th Most Common Venue   8th Most Common Venue  \
0                    Wine Bar              Print Shop   
1                        Café            Soccer Field   
2        Fast Food Restaurant       Convenience Store   
3        Fast Food Restaurant       Convenience Store   
4          Italian Restaurant                  Bakery   
5                Burger Joint                Bus Stop   
6                 Supermarket                    Park   
7                        Park                Bus Stop   
8                        Café      Light Rail Station   
9                 Coffee Shop                    Park   
10                Flea Market         Thai Restaurant   
11  Middle Eastern Restaurant  Thrift / Vintage Store   
12             Sandwich Place           Movie Theater   
13               Burger Joint                 Brewery   
14                       Café    Fast Food Restaurant   

In [29]:
BROM_neighborhoods_venues_sorted

Neighbourhood    1st Most Common Venue  \
0                          Bickley           Cricket Ground   
1                      Biggin Hill                  Airport   
2          Bromley Common & Keston                      Pub   
3                     Bromley Town              Coffee Shop   
4       Chelsfield & Pratts Bottom  Health & Beauty Service   
5                      Chislehurst                      Pub   
6                      Copers Cope       Athletics & Sports   
7                 Cray Valley East              Coffee Shop   
8                   Crystal Palace                 Platform   
9                           Darwin           History Museum   
10          Farnborough & Croftton                      Pub   
11              Hayes & Coney Hall                     Park   
12              Kelsey & Eden Park                     Café   
13  Mottingham & Chislehurst North              Pizza Place   
14                       Orpington     Fast Food Restaurant   
15                   Penge & Cator                      Pub   
16              Petts Wood & Knoll        Convenience Store   
17           Plaistow & Sunbridge         Other Repair Shop   
18                     Shortlangs               Golf Course   
19                   West Wickham                      Park   

   2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0          Train Station                    Café          Home Service   
1        Airport Service                     Pub        Massage Studio   
2                    Bar       Indian Restaurant           Bus Station   
3         Clothing Store    Gym / Fitness Center                   Pub   
4                    Pub           Fishing Store         Women's Store   
5     Italian Restaurant               Gastropub           Pizza Place   
6           Soccer Field        Indoor Play Area        Cricket Ground   
7         Clothing Store    Gym / Fitness Center        Sandwich Place   
8         Breakfast Spot                    Park      Sculpture Garden   
9                    Bar                    Farm         Women's Store   
10   American Restaurant                 Theater           Coffee Shop   
11         Grocery Store          Hardware Store                   Pub   
12          Tennis Court                    Park         Women's Store   
13     Fish & Chips Shop                     Pub         Grocery Store   
14    Italian Restaurant             Supermarket    Chinese Restaurant   
15           Supermarket             Pizza Place                  Park   
16             Gift Shop            Home Service                  Farm   
17    Chinese Restaurant                   Diner           Supermarket   
18         Train Station      Chinese Restaurant                  Café   
19      Business Service  Furniture / Home Store                  Farm   

   5th Most Common Venue   6th Most Common Venue   7th Most Common Venue  \
0          Women's Store    Fast Food Restaurant        Department Store   
1                   Farm          Cricket Ground        Department Store   
2   Fast Food Restaurant           Women's Store                    Farm   
3            Pizza Place            Burger Joint        Department Store   
4                   Farm          Cosmetics Shop          Cricket Ground   
5      Indian Restaurant           Women's Store          Cosmetics Shop   
6       Department Store                   Diner              Donut Shop   
7              Gastropub                    Café          Ice Cream Shop   
8              Wine Shop    Gym / Fitness Center                    Farm   
9   Fast Food Restaurant          Cricket Ground        Department Store   
10         Women's Store                    Farm          Cosmetics Shop   
11                  Farm          Cosmetics Shop          Cricket Ground   
12                  Farm          Cosmetics Shop          Cricket Ground   
13         Garden Center                  Garden  Furniture / Home Store   
14

## Cluster Neighbourhoods

In [30]:
# Run K-Mean to cluster the neighbourhoods in each borough into 5 clusters 
TH_BROM_venues = TH_venues.merge(BROM_venues, how='outer')
TH_BROM_venues

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0            Bethnal Green               51.526962                -0.066740   
1            Bethnal Green               51.526962                -0.066740   
2            Bethnal Green               51.526962                -0.066740   
3            Bethnal Green               51.526962                -0.066740   
4            Bethnal Green               51.526962                -0.066740   
5            Bethnal Green               51.526962                -0.066740   
6            Bethnal Green               51.526962                -0.066740   
7            Bethnal Green               51.526962                -0.066740   
8            Bethnal Green               51.526962                -0.066740   
9            Bethnal Green               51.526962                -0.066740   
10           Bethnal Green               51.526962                -0.066740   
11           Bethnal Green               51.526962                -0.066740   
12           Bethnal Green               51.526962                -0.066740   
13           Bethnal Green               51.526962                -0.066740   
14           Bethnal Green               51.526962                -0.066740   
15           Bethnal Green               51.526962                -0.066740   
16           Bethnal Green               51.526962                -0.066740   
17           Bethnal Green               51.526962                -0.066740   
18           Bethnal Green               51.526962                -0.066740   
19           Bethnal Green               51.526962                -0.066740   
20           Bethnal Green               51.526962                -0.066740   
21           Bethnal Green               51.526962                -0.066740   
22           Bethnal Green               51.526962                -0.066740   
23           Bethnal Green               51.526962                -0.066740   
24           Bethnal Green               51.526962                -0.066740   
25           Bethnal Green               51.526962                -0.066740   
26           Bethnal Green               51.526962                -0.066740   
27           Bethnal Green               51.526962                -0.066740   
28           Bethnal Green               51.526962                -0.066740   
29           Bethnal Green               51.526962                -0.066740   
..                     ...                     ...                      ...   
683          Penge & Cator               51.415205                -0.053734   
684          Penge & Cator               51.415205                -0.053734   
685          Penge & Cator               51.415205                -0.053734   
686          Penge & Cator               51.415205                -0.053734   
687          Penge & Cator               51.415205                -0.053734   
688          Penge & Cator               51.415205                -0.053734   
689          Penge & Cator               51.415205                -0.053734   
690          Penge & Cator               51.415205                -0.053734   
691          Penge & Cator               51.415205                -0.053734   
692          Penge & Cator               51.415205                -0.053734   
693          Penge & Cator               51.415205                -0.053734   
694          Penge & Cator               51.415205                -0.053734   
695          Penge & Cator               51.415205                -0.053734   
696          Penge & Cator               51.415205                -0.053734   
697          Penge & Cator               51.415205                -0.053734   
698     Petts Wood & Knoll               51.386370                 0.064750   
699     Petts Wood & Knoll               51.386370                 0.064750   
700     Petts Wood & Knoll               51.386370                 0.064750   
701  Plaistow & Sunbridge                51.419800             

In [31]:
# one hot encoding
TH_BROM_onehot = pd.get_dummies(TH_BROM_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
TH_BROM_onehot['Neighbourhood'] = TH_BROM_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [TH_BROM_onehot.columns[-1]] + list(TH_BROM_onehot.columns[:-1])
TH_BROM_onehot = TH_BROM_onehot[fixed_columns]

TH_BROM_grouped = TH_BROM_onehot.groupby('Neighbourhood').mean().reset_index()
TH_BROM_grouped

Neighbourhood  Airport  Airport Service  \
0                    Bethnal Green     0.00             0.00   
1                          Bickley     0.00             0.00   
2                      Biggin Hill     0.25             0.25   
3        Blackwall and Cubitt Town     0.00             0.00   
4                         Bow East     0.00             0.00   
5                         Bow West     0.00             0.00   
6          Bromley Common & Keston     0.00             0.00   
7                     Bromley Town     0.00             0.00   
8                     Canary Wharf     0.00             0.00   
9       Chelsfield & Pratts Bottom     0.00             0.00   
10                     Chislehurst     0.00             0.00   
11                     Copers Cope     0.00             0.00   
12                Cray Valley East     0.00             0.00   
13                  Crystal Palace     0.00             0.00   
14                          Darwin     0.00             0.00   
15          Farnborough & Croftton     0.00             0.00   
16              Hayes & Coney Hall     0.00             0.00   
17                 Island Gardens      0.00             0.00   
18              Kelsey & Eden Park     0.00             0.00   
19                        Lansbury     0.00             0.00   
20                      Limehouse      0.00             0.00   
21  Mottingham & Chislehurst North     0.00             0.00   
22                       Orpington     0.00             0.00   
23                   Penge & Cator     0.00             0.00   
24              Petts Wood & Knoll     0.00             0.00   
25           Plaistow & Sunbridge      0.00             0.00   
26                          Poplar     0.00             0.00   
27                        Shadwell     0.00             0.00   
28                     Shortlangs      0.00             0.00   
29     Spitalfields and Banglatown     0.00             0.00   
30     St Katharine's and Wapping      0.00             0.00   
31                   Stepney Green     0.00             0.00   
32                         Weavers     0.00             0.00   
33                   West Wickham      0.00             0.00   
34                     Whitechapel     0.00             0.00   

    American Restaurant  Art Gallery  Asian Restaurant  Athletics & Sports  \
0                   0.0     0.015873          0.000000            0.000000   
1                   0.0     0.000000          0.000000            0.000000   
2                   0.0     0.000000          0.000000            0.000000   
3                   0.0     0.000000          0.000000            0.000000   
4                   0.0     0.076923          0.000000            0.000000   
5                   0.0     0.076923          0.000000            0.000000   
6                   0.0     0.000000          0.000000            0.000000   
7                   0.0     0.000000          0.022222            0.000000   
8                   0.0     0.000000          0.010000            0.000000   
9                   0.0     0.000000          0.000000            0.000000   
10                  0.0     0.000000          0.000000            0.000000   
11                  0.0     0.000000          0.000000            0.500000   
12                  0.0     0.000000          0.035714            0.000000   
13                  0.0     0.000000          0.000000            0.045455   
14                  0.0     0.000000          0.000000            0.000000   
15                  0.2     0.000000          0.000000            0.000000   
16                  0.0     0.000000          0.000000            0.000000   
17                  0.0     0.033333          0.000000            0.033333   
18                  0.0     0.000000          0.000000            0.000000   
19                  0.0     0.000000          0.062500            0.000000   
20                  0.0     0.000000          0.000000            0.035714   


In [32]:
num_top_venues = 5

for hood in TH_BROM_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = TH_BROM_grouped[TH_BROM_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bethnal Green----
         venue  freq
0  Coffee Shop  0.13
1          Pub  0.11
2         Café  0.06
3     Wine Bar  0.03
4   Bagel Shop  0.03


----Bickley----
            venue  freq
0            Café  0.25
1  Cricket Ground  0.25
2    Home Service  0.25
3   Train Station  0.25
4         Airport  0.00


----Biggin Hill----
             venue  freq
0          Airport  0.25
1              Pub  0.25
2   Massage Studio  0.25
3  Airport Service  0.25
4        Wine Shop  0.00


----Blackwall and Cubitt Town----
               venue  freq
0  Indian Restaurant  0.09
1               Park  0.09
2               Café  0.06
3                Pub  0.06
4        Pizza Place  0.06


----Bow East----
                      venue  freq
0                       Pub  0.15
1      Fast Food Restaurant  0.08
2             Metro Station  0.08
3                       Gym  0.08
4  Bike Rental / Bike Share  0.08


----Bow West----
                      venue  freq
0                       Pub  0.15
1      Fas

In [33]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
TH_BROM_neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
TH_BROM_neighborhoods_venues_sorted['Neighbourhood'] = TH_BROM_grouped['Neighbourhood']

for ind in np.arange(TH_BROM_grouped.shape[0]):
    TH_BROM_neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(TH_BROM_grouped.iloc[ind, :], num_top_venues)

In [34]:
TH_BROM_neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                    Bethnal Green           Coffee Shop   
1                          Bickley          Home Service   
2                      Biggin Hill               Airport   
3        Blackwall and Cubitt Town     Indian Restaurant   
4                         Bow East                   Pub   
5                         Bow West                   Pub   
6          Bromley Common & Keston           Bus Station   
7                     Bromley Town           Coffee Shop   
8                     Canary Wharf           Coffee Shop   
9       Chelsfield & Pratts Bottom                   Pub   
10                     Chislehurst                   Pub   
11                     Copers Cope    Athletics & Sports   
12                Cray Valley East           Coffee Shop   
13                  Crystal Palace              Platform   
14                          Darwin                  Farm   
15          Farnborough & Croftton                   Pub   
16              Hayes & Coney Hall                  Park   
17                 Island Gardens                    Pub   
18              Kelsey & Eden Park                  Café   
19                        Lansbury     Indian Restaurant   
20                      Limehouse                    Pub   
21  Mottingham & Chislehurst North                   Pub   
22                       Orpington  Fast Food Restaurant   
23                   Penge & Cator                   Pub   
24              Petts Wood & Knoll          Home Service   
25           Plaistow & Sunbridge     Chinese Restaurant   
26                          Poplar                  Park   
27                        Shadwell                 Hotel   
28                     Shortlangs     Chinese Restaurant   
29     Spitalfields and Banglatown           Coffee Shop   
30     St Katharine's and Wapping                    Pub   
31                   Stepney Green                  Park   
32                         Weavers                   Pub   
33                   West Wickham                   Park   
34                     Whitechapel                   Pub   

      2nd Most Common Venue          3rd Most Common Venue  \
0                       Pub                           Café   
1            Cricket Ground                           Café   
2           Airport Service                            Pub   
3                      Park                Harbor / Marina   
4                     Hotel                            Bar   
5                     Hotel                            Bar   
6         Indian Restaurant                            Pub   
7            Clothing Store           Gym / Fitness Center   
8            Sandwich Place                   Burger Joint   
9   Health & Beauty Service                  Fishing Store   
10        Indian Restaurant                    Pizza Place   
11             Soccer Field               Indoor Play Area   
12           Clothing Store           Gym / Fitness Center   
13           Breakfast Spot                 History Museum   
14           History Museum                            Bar   
15      American Restaurant                    Coffee Shop   
16                      Pub                 Hardware Store   
17           History Museum                  Boat or Ferry   
18             Tennis Court                           Park   
19               Playground                  Grocery Store   
20       Chinese Restaurant                           Café   
21              Pizza Place              Fish & Chips Shop   
22              Supermarket             Italian Restaurant   
23              Pizza Place                    Supermarket   
24        Convenience Store                      Gift Shop   
25                    Diner                    Supermarket   
26               Steakhouse                    Pizza Place   
27            Grocery Store                            Pub   
28                      Pub                           Café   


In [35]:
# set number of clusters
kclusters = 5

TH_BROM_grouped_clustering = TH_BROM_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(TH_BROM_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 4, 2, 4, 4, 4, 2, 2, 4], dtype=int32)

In [42]:
TH_BROM_neighborhoods_venues_sorted

Cluster Label                   Neighbourhood 1st Most Common Venue  \
0               2                   Bethnal Green           Coffee Shop   
1               2                         Bickley          Home Service   
2               4                     Biggin Hill               Airport   
3               2       Blackwall and Cubitt Town     Indian Restaurant   
4               4                        Bow East                   Pub   
5               4                        Bow West                   Pub   
6               4         Bromley Common & Keston           Bus Station   
7               2                    Bromley Town           Coffee Shop   
8               2                    Canary Wharf           Coffee Shop   
9               4      Chelsfield & Pratts Bottom                   Pub   
10              4                     Chislehurst                   Pub   
11              3                     Copers Cope    Athletics & Sports   
12              2                Cray Valley East           Coffee Shop   
13              2                  Crystal Palace              Platform   
14              1                          Darwin                  Farm   
15              4          Farnborough & Croftton                   Pub   
16              0              Hayes & Coney Hall                  Park   
17              4                 Island Gardens                    Pub   
18              0              Kelsey & Eden Park                  Café   
19              2                        Lansbury     Indian Restaurant   
20              4                      Limehouse                    Pub   
21              4  Mottingham & Chislehurst North                   Pub   
22              2                       Orpington  Fast Food Restaurant   
23              4                   Penge & Cator                   Pub   
24              2              Petts Wood & Knoll          Home Service   
25              0           Plaistow & Sunbridge     Chinese Restaurant   
26              0                          Poplar                  Park   
27              2                        Shadwell                 Hotel   
28              4                     Shortlangs     Chinese Restaurant   
29              2     Spitalfields and Banglatown           Coffee Shop   
30              4     St Katharine's and Wapping                    Pub   
31              0                   Stepney Green                  Park   
32              4                         Weavers                   Pub   
33              0                   West Wickham                   Park   
34              2                     Whitechapel                   Pub   

      2nd Most Common Venue          3rd Most Common Venue  \
0                       Pub                           Café   
1            Cricket Ground                           Café   
2           Airport Service                            Pub   
3                      Park                Harbor / Marina   
4                     Hotel                            Bar   
5                     Hotel                            Bar   
6         Indian Restaurant                            Pub   
7            Clothing Store           Gym / Fitness Center   
8            Sandwich Place                   Burger Joint   
9   Health & Beauty Service                  Fishing Store   
10        Indian Restaurant                    Pizza Place   
11             Soccer Field               Indoor Play Area   
12           Clothing Store           Gym / Fitness Center   
13           Breakfast Spot                 History Museum   
14           History Museum                            Bar   
15      American Restaurant                    Coffee Shop   
16                      Pub                 Hardware Store   
17           History Museum                  Boat or Ferry   
18             Tennis Court                           Park   
19               Playground                  Grocery Store   
2

========================================================================================================================

## <span style="color:darkred">5. Results and Discussions <a name="results"></a></span>

In [36]:
print(TH_neighborhoods_venues_sorted['1st Most Common Venue'].value_counts().head(3))
print(TH_neighborhoods_venues_sorted['2nd Most Common Venue'].value_counts().head(3))
print(TH_neighborhoods_venues_sorted['3rd Most Common Venue'].value_counts().head(3))

Pub                  7
Coffee Shop          3
Indian Restaurant    2
Name: 1st Most Common Venue, dtype: int64
Pub                         3
Hotel                       2
Bike Rental / Bike Share    2
Name: 2nd Most Common Venue, dtype: int64
Hotel             2
Sandwich Place    2
Boat or Ferry     1
Name: 3rd Most Common Venue, dtype: int64


In [37]:
print(BROM_neighborhoods_venues_sorted['1st Most Common Venue'].value_counts().head(3))
print(BROM_neighborhoods_venues_sorted['2nd Most Common Venue'].value_counts().head(3))
print(BROM_neighborhoods_venues_sorted['3rd Most Common Venue'].value_counts().head(3))

Pub            4
Coffee Shop    2
Park           2
Name: 1st Most Common Venue, dtype: int64
Italian Restaurant    2
Train Station         2
Clothing Store        2
Name: 2nd Most Common Venue, dtype: int64
Park                    2
Gym / Fitness Center    2
Pub                     2
Name: 3rd Most Common Venue, dtype: int64


### Analysis of the most common venues 

Let's analyse the most common venue between the two boroughs:
* Tower Hamlet: refering to the first 3 most common venues in Tower Hamlet, we can easily identify that mist venues are food and beverage venues, with 84% of the venues. It is seems that in most of the Tower Hamlet venues are for essential sustenance providers. Furthermore, due to the location of the borough, the boroughs has provided bike rental and bike share venues for the residents to commute within the area.  
* Bromley: However, in Bromley, althougb also includes food and beverage venues, the venues also includes clothing stores, gyms, parks and Italian restaurants, as such that these are life style luxuries that are mostly common for area that have low poverty rates. 

If we looking at the big pictures between the two boroughs, we can somehow identity that in general, majority of the venues in Tower Hamlets are food and beverage or sustenance venues, such as cafe, pubs restaurant, grocery stores etc. However, Bromley provides more lifestyle venues such as Gyms, massage studios, health and beuty studio, golf course, theater and garden center just to name a few. 

In summary, currently, there are clear discrepancies between venues in these two boroughs based on the poverty rate in these two boroughs. 

Now that we know the dicrepancies between the two boroughs, are there any similarities between two boroughs?

### Examine Clusters

#### Cluster 1:

In [51]:
TH_BROM_neighborhoods_venues_sorted.loc[TH_BROM_neighborhoods_venues_sorted['Cluster Label'] == 0]

Cluster Label          Neighbourhood 1st Most Common Venue  \
16              0     Hayes & Coney Hall                  Park   
18              0     Kelsey & Eden Park                  Café   
25              0  Plaistow & Sunbridge     Chinese Restaurant   
26              0                 Poplar                  Park   
31              0          Stepney Green                  Park   
33              0          West Wickham                   Park   

     2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
16                     Pub        Hardware Store         Grocery Store   
18            Tennis Court                  Park           Yoga Studio   
25                   Diner           Supermarket     Other Repair Shop   
26              Steakhouse           Pizza Place    Light Rail Station   
31      Chinese Restaurant                   Pub          Dessert Shop   
33  Furniture / Home Store      Business Service            Food Truck   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
16             Yoga Studio    Falafel Restaurant           Flower Shop   
18    Fast Food Restaurant            Food Court     Food & Drink Shop   
25    Fast Food Restaurant            Food Court     Food & Drink Shop   
26                    Café   Fried Chicken Joint          Intersection   
31  Thrift / Vintage Store        Sandwich Place   Fried Chicken Joint   
33              Food Court     Food & Drink Shop           Flower Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
16           Flea Market         Fishing Store            Fish Market  
18           Flower Shop           Flea Market          Fishing Store  
25           Flower Shop           Flea Market          Fishing Store  
26                Tunnel    Chinese Restaurant     English Restaurant  
31                  Farm           Supermarket     Italian Restaurant  
33           Flea Market         Fishing Store            Fish Market

Let's examine cluster 1. Based on the data from boith boroughs, in cluster 1, only poplar is located at Tower Hamlet. Further investigation indicates thatpopplar is located at the boundary of the finacial hub of London, canary wharf, where most financial sector employees are resided with low poverty rate. Due to this reason, the venues are similar with the venues in cluster in Bromley. 

#### Cluster 2:

In [53]:
TH_BROM_neighborhoods_venues_sorted.loc[TH_BROM_neighborhoods_venues_sorted['Cluster Label'] == 1]

Cluster Label Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
14              1        Darwin                  Farm        History Museum   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
14                   Bar           Yoga Studio            Food Stand   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
14            Food Court     Food & Drink Shop           Flower Shop   

   9th Most Common Venue 10th Most Common Venue  
14           Flea Market          Fishing Store

Cluster 2 only have one neighbourhood, which is Darwin, located in Bromley. Analysing the venues showsn that the area are leisure areas where more lifestyle venues are located, such as yoga studio, flower shop, bars etc. 

#### Cluster 3:

In [54]:
TH_BROM_neighborhoods_venues_sorted.loc[TH_BROM_neighborhoods_venues_sorted['Cluster Label'] == 2]

Cluster Label                Neighbourhood 1st Most Common Venue  \
0               2                Bethnal Green           Coffee Shop   
1               2                      Bickley          Home Service   
3               2    Blackwall and Cubitt Town     Indian Restaurant   
7               2                 Bromley Town           Coffee Shop   
8               2                 Canary Wharf           Coffee Shop   
12              2             Cray Valley East           Coffee Shop   
13              2               Crystal Palace              Platform   
19              2                     Lansbury     Indian Restaurant   
22              2                    Orpington  Fast Food Restaurant   
24              2           Petts Wood & Knoll          Home Service   
27              2                     Shadwell                 Hotel   
29              2  Spitalfields and Banglatown           Coffee Shop   
34              2                  Whitechapel                   Pub   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0                    Pub                           Café              Wine Bar   
1         Cricket Ground                           Café         Train Station   
3                   Park                Harbor / Marina           Pizza Place   
7         Clothing Store           Gym / Fitness Center                   Pub   
8         Sandwich Place                   Burger Joint                 Plaza   
12        Clothing Store           Gym / Fitness Center        Chocolate Shop   
13        Breakfast Spot                 History Museum         Track Stadium   
19            Playground                  Grocery Store                  Café   
22           Supermarket             Italian Restaurant            Restaurant   
24     Convenience Store                      Gift Shop           Yoga Studio   
27         Grocery Store                            Pub    Italian Restaurant   
29                   Pub  Vegetarian / Vegan Restaurant            Food Truck   
34     Indian Restaurant                 Sandwich Place           Coffee Shop   

   5th Most Common Venue 6th Most Common Venue   7th Most Common Venue  \
0                   Park            Bagel Shop              Restaurant   
1            Yoga Studio                  Farm              Food Court   
3          Grocery Store                  Café                     Pub   
7           Burger Joint           Pizza Place           Movie Theater   
8             Food Truck                Bakery      Italian Restaurant   
12             Bookstore      Stationery Store  Furniture / Home Store   
13           Bus Station      Sculpture Garden                  Garden   
19                 Canal  Fast Food Restaurant             Supermarket   
22                  Café           Coffee Shop      Chinese Restaurant   
24                  Farm            Food Court       Food & Drink Shop   
27           Pizza Place          Burger Joint    Fast Food Restaurant   
29     Indian Restaurant           Flea Market             Art Gallery   
34                 Hotel        Ice Cream Shop             Supermarket   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Shoe Store            Print Shop          Jewelry Store  
1      Food & Drink Shop           Flower Shop            Flea Market  
3             Restaurant    Light Rail Station               Bus Stop  
7          Burrito Place                  Park                   Café  
8          Shopping Mall                  Café       Sushi Restaurant  
12         Movie Theater                  Café         Sandwich Place  
13                  Farm             Bike Shop                   Park  
19  Pakistani Restaurant     Food & Drink Shop            Beer Garden  
22         Memorial Site                Bakery  Portuguese Restaurant  
24           Flower Shop           Flea Market          Fishing Store  
27                Market       

#### Cluster 4:

In [55]:
TH_BROM_neighborhoods_venues_sorted.loc[TH_BROM_neighborhoods_venues_sorted['Cluster Label'] == 3]

Cluster Label Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
11              3   Copers Cope    Athletics & Sports          Soccer Field   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11      Indoor Play Area           Yoga Studio                  Farm   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11            Food Court     Food & Drink Shop           Flower Shop   

   9th Most Common Venue 10th Most Common Venue  
11           Flea Market          Fishing Store

Cluster 4 only have one neighbourhood, which is Copers , located in Bromley. Analysing the venues showsn that the area are leisure areas where more sport lifestyle venues are located, such as atheletics & sports, soccer field, indoor play area and yoga studio. 

#### Cluster 5:

In [52]:
TH_BROM_neighborhoods_venues_sorted.loc[TH_BROM_neighborhoods_venues_sorted['Cluster Label'] == 4]

Cluster Label                   Neighbourhood 1st Most Common Venue  \
2               4                     Biggin Hill               Airport   
4               4                        Bow East                   Pub   
5               4                        Bow West                   Pub   
6               4         Bromley Common & Keston           Bus Station   
9               4      Chelsfield & Pratts Bottom                   Pub   
10              4                     Chislehurst                   Pub   
15              4          Farnborough & Croftton                   Pub   
17              4                 Island Gardens                    Pub   
20              4                      Limehouse                    Pub   
21              4  Mottingham & Chislehurst North                   Pub   
23              4                   Penge & Cator                   Pub   
28              4                     Shortlangs     Chinese Restaurant   
30              4     St Katharine's and Wapping                    Pub   
32              4                         Weavers                   Pub   

      2nd Most Common Venue      3rd Most Common Venue  4th Most Common Venue  \
2           Airport Service                        Pub         Massage Studio   
4                     Hotel                        Bar           Burger Joint   
5                     Hotel                        Bar           Burger Joint   
6         Indian Restaurant                        Pub                    Bar   
9   Health & Beauty Service              Fishing Store            Yoga Studio   
10        Indian Restaurant                Pizza Place     Italian Restaurant   
15      American Restaurant                Coffee Shop                Theater   
17           History Museum              Boat or Ferry                   Park   
20       Chinese Restaurant                       Café      Indian Restaurant   
21              Pizza Place          Fish & Chips Shop          Grocery Store   
23              Pizza Place                Supermarket      Convenience Store   
28                      Pub                       Café          Train Station   
30                 Bus Stop  Middle Eastern Restaurant         Sandwich Place   
32              Supermarket             Ice Cream Shop  Portuguese Restaurant   

    5th Most Common Venue   6th Most Common Venue   7th Most Common Venue  \
2                    Farm              Food Court       Food & Drink Shop   
4           Metro Station                  Office    Fast Food Restaurant   
5           Metro Station                  Office    Fast Food Restaurant   
6    Fast Food Restaurant                    Farm              Food Stand   
9      Falafel Restaurant              Food Court       Food & Drink Shop   
10              Gastropub             Yoga Studio                    Farm   
15            Yoga Studio                    Farm              Food Court   
17  Portuguese Restaurant                  Church              Restaurant   
20     Italian Restaurant                    Park                Bus Stop   
21            Yoga Studio      Falafel Restaurant       Food & Drink Shop   
23          Shopping Mall    Fast Food Restaurant  Furniture / Home Store   
28            Golf Course             Yoga Studio       Food & Drink Shop   
30            Coffee Shop  Thrift / Vintage Store                     Bar   
32           Burger Joint                 Brewery          Sandwich Place   

       8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
2                Flower Shop           Flea Market           Fishing Store  
4   Bike Rental / Bike Share     Convenience Store                     Gym  
5   Bike Rental / Bike Share     Convenience Store                     Gym  
6                 Food Court     Food & Drink Shop             Flower Shop  
9                Flower Shop           Flea Market             Fish Market  
10         Food & Drink Shop           Flowe

Finally, cluster 5 have even split between two boroughs. Generally, this cluster consists of the food and drinks venues in both boroughs. From this cluster, we can clearly identify that pubs are the favourite leisure venues for all residents in both areas, regardless of the poverty rate. 

======================================================================================================================== 

## <span style="color:darkred">6. Conclusions <a name="conclusion"></a></span>

In conclusion, based on the poverty rate, we can clearly see there are discrepancies in terms of the venues at each boroughs. At Tower Hamlet, it is clearly shown that most venues provided are essentail venues for day-to-day life as in Bromley , much of the venues are lifestyle venues and required certain level of incomes. 

======================================================================================================================== 